File: src/bun.js/test/expect/toIncludeRepeated.zig
pub fn toIncludeRepeated(this: *Expect, globalThis: *JSGlobalObject, callFrame: *CallFrame) bun.JSError!JSValue {
    defer this.postMatch(globalThis);

    const thisValue = callFrame.this();
    const arguments_ = callFrame.arguments_old(2);
    const arguments = arguments_.slice();

    if (arguments.len < 2) {
        return globalThis.throwInvalidArguments("toIncludeRepeated() requires 2 arguments", .{});
    }

    this.incrementExpectCallCounter();

    const substring = arguments[0];
    substring.ensureStillAlive();

    if (!substring.isString()) {
        return globalThis.throw("toIncludeRepeated() requires the first argument to be a string", .{});
    }

    const count = arguments[1];
    count.ensureStillAlive();

    if (!count.isAnyInt()) {
        return globalThis.throw("toIncludeRepeated() requires the second argument to be a number", .{});
    }

    const countAsNum = count.toU32();

    const expect_string = Expect.js.capturedValueGetCached(thisValue) orelse {
        return globalThis.throw("Internal consistency error: the expect(value) was garbage collected but it should not have been!", .{});
    };

    if (!expect_string.isString()) {
        return globalThis.throw("toIncludeRepeated() requires the expect(value) to be a string", .{});
    }

    const not = this.flags.not;
    var pass = false;

    const _expectStringAsStr = try expect_string.toSliceOrNull(globalThis);
    const _subStringAsStr = try substring.toSliceOrNull(globalThis);

    defer {
        _expectStringAsStr.deinit();
        _subStringAsStr.deinit();
    }

    const expectStringAsStr = _expectStringAsStr.slice();
    const subStringAsStr = _subStringAsStr.slice();

    if (subStringAsStr.len == 0) {
        return globalThis.throw("toIncludeRepeated() requires the first argument to be a non-empty string", .{});
    }

    const actual_count = std.mem.count(u8, expectStringAsStr, subStringAsStr);
    pass = actual_count == countAsNum;

    if (not) pass = !pass;
    if (pass) return .js_undefined;

    var formatter = jsc.ConsoleObject.Formatter{ .globalThis = globalThis, .quote_strings = true };
    defer formatter.deinit();
    const expect_string_fmt = expect_string.toFmt(&formatter);
    const substring_fmt = substring.toFmt(&formatter);
    const times_fmt = count.toFmt(&formatter);

    const received_line = "Received: <red>{f}<r>\n";

    if (not) {
        if (countAsNum == 0) {
            const expected_line = "Expected to include: <green>{f}<r> \n";
            const signature = comptime getSignature("toIncludeRepeated", "<green>expected<r>", true);
            return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ substring_fmt, expect_string_fmt });
        } else if (countAsNum == 1) {
            const expected_line = "Expected not to include: <green>{f}<r> \n";
            const signature = comptime getSignature("toIncludeRepeated", "<green>expected<r>", true);
            return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ substring_fmt, expect_string_fmt });
        } else {
            const expected_line = "Expected not to include: <green>{f}<r> <green>{f}<r> times \n";
            const signature = comptime getSignature("toIncludeRepeated", "<green>expected<r>", true);
            return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ substring_fmt, times_fmt, expect_string_fmt });
        }
    }

    if (countAsNum == 0) {
        const expected_line = "Expected to not include: <green>{f}<r>\n";
        const signature = comptime getSignature("toIncludeRepeated", "<green>expected<r>", false);
        return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ substring_fmt, expect_string_fmt });
    } else if (countAsNum == 1) {
        const expected_line = "Expected to include: <green>{f}<r>\n";
        const signature = comptime getSignature("toIncludeRepeated", "<green>expected<r>", false);
        return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ substring_fmt, expect_string_fmt });
    } else {
        const expected_line = "Expected to include: <green>{f}<r> <green>{f}<r> times \n";
        const signature = comptime getSignature("toIncludeRepeated", "<green>expected<r>", false);
        return this.throw(globalThis, signature, "\n\n" ++ expected_line ++ received_line, .{ substring_fmt, times_fmt, expect_string_fmt });
    }
}

const bun = @import("bun");
const std = @import("std");

const jsc = bun.jsc;
const CallFrame = bun.jsc.CallFrame;
const JSGlobalObject = bun.jsc.JSGlobalObject;
const JSValue = bun.jsc.JSValue;

const Expect = bun.jsc.Expect.Expect;
const getSignature = Expect.getSignature;

File: src/test_runner/expect/toIncludeRepeated.rs
use bstr::ByteSlice;
use bun_jsc::{CallFrame, JSGlobalObject, JSValue, JsResult};
use bun_jsc::console_object::Formatter;

use super::{Expect, get_signature};

impl Expect {
    #[bun_jsc::host_fn(method)]
    pub fn to_include_repeated(
        this: &mut Self,
        global: &JSGlobalObject,
        frame: &CallFrame,
    ) -> JsResult<JSValue> {
        // TODO(port): `defer this.postMatch(global)` — scopeguard captures &mut Self and
        // conflicts with later uses; Phase B may need an inner-fn + post_match-on-exit reshape.
        let _post = scopeguard::guard((), |_| this.post_match(global));
        // PORT NOTE: reshaped for borrowck (see above)

        let this_value = frame.this();
        let arguments_ = frame.arguments_old(2);
        let arguments = arguments_.slice();

        if arguments.len() < 2 {
            return global.throw_invalid_arguments(format_args!(
                "toIncludeRepeated() requires 2 arguments"
            ));
        }

        this.increment_expect_call_counter();

        let substring = arguments[0];
        substring.ensure_still_alive();

        if !substring.is_string() {
            return global.throw(format_args!(
                "toIncludeRepeated() requires the first argument to be a string"
            ));
        }

        let count = arguments[1];
        count.ensure_still_alive();

        if !count.is_any_int() {
            return global.throw(format_args!(
                "toIncludeRepeated() requires the second argument to be a number"
            ));
        }

        let count_as_num = count.to_u32();

        let Some(expect_string) = Expect::js::captured_value_get_cached(this_value) else {
            return global.throw(format_args!(
                "Internal consistency error: the expect(value) was garbage collected but it should not have been!"
            ));
        };

        if !expect_string.is_string() {
            return global.throw(format_args!(
                "toIncludeRepeated() requires the expect(value) to be a string"
            ));
        }

        let not = this.flags.not;
        let mut pass = false;

        let expect_string_as_str_owned = expect_string.to_slice_or_null(global)?;
        let sub_string_as_str_owned = substring.to_slice_or_null(global)?;
        // defer .deinit() → handled by Drop

        let expect_string_as_str = expect_string_as_str_owned.slice();
        let sub_string_as_str = sub_string_as_str_owned.slice();

        if sub_string_as_str.is_empty() {
            return global.throw(format_args!(
                "toIncludeRepeated() requires the first argument to be a non-empty string"
            ));
        }

        // std.mem.count(u8, haystack, needle) — non-overlapping occurrence count
        let actual_count = expect_string_as_str.find_iter(sub_string_as_str).count();
        pass = actual_count == count_as_num as usize;

        if not {
            pass = !pass;
        }
        if pass {
            return Ok(JSValue::UNDEFINED);
        }

        let mut formatter = Formatter {
            global_this: global,
            quote_strings: true,
            ..Default::default()
        };
        // defer formatter.deinit() → handled by Drop
        // PORT NOTE: to_fmt borrows the formatter; three live borrows below would alias under
        // &mut. Using shared & here; Phase B may need interior mutability on Formatter.
        let expect_string_fmt = expect_string.to_fmt(&formatter);
        let substring_fmt = substring.to_fmt(&formatter);
        let times_fmt = count.to_fmt(&formatter);

        const RECEIVED_LINE: &str = "Received: <red>{}<r>\n";

        if not {
            if count_as_num == 0 {
                const EXPECTED_LINE: &str = "Expected to include: <green>{}<r> \n";
                const SIGNATURE: &str = get_signature("toIncludeRepeated", "<green>expected<r>", true);
                return this.throw(
                    global,
                    SIGNATURE,
                    format_args!(
                        concat!("\n\n", "Expected to include: <green>{}<r> \n", "Received: <red>{}<r>\n"),
                        substring_fmt,
                        expect_string_fmt
                    ),
                );
            } else if count_as_num == 1 {
                const EXPECTED_LINE: &str = "Expected not to include: <green>{}<r> \n";
                const SIGNATURE: &str = get_signature("toIncludeRepeated", "<green>expected<r>", true);
                return this.throw(
                    global,
                    SIGNATURE,
                    format_args!(
                        concat!("\n\n", "Expected not to include: <green>{}<r> \n", "Received: <red>{}<r>\n"),
                        substring_fmt,
                        expect_string_fmt
                    ),
                );
            } else {
                const EXPECTED_LINE: &str = "Expected not to include: <green>{}<r> <green>{}<r> times \n";
                const SIGNATURE: &str = get_signature("toIncludeRepeated", "<green>expected<r>", true);
                return this.throw(
                    global,
                    SIGNATURE,
                    format_args!(
                        concat!("\n\n", "Expected not to include: <green>{}<r> <green>{}<r> times \n", "Received: <red>{}<r>\n"),
                        substring_fmt,
                        times_fmt,
                        expect_string_fmt
                    ),
                );
            }
        }

        if count_as_num == 0 {
            const EXPECTED_LINE: &str = "Expected to not include: <green>{}<r>\n";
            const SIGNATURE: &str = get_signature("toIncludeRepeated", "<green>expected<r>", false);
            this.throw(
                global,
                SIGNATURE,
                format_args!(
                    concat!("\n\n", "Expected to not include: <green>{}<r>\n", "Received: <red>{}<r>\n"),
                    substring_fmt,
                    expect_string_fmt
                ),
            )
        } else if count_as_num == 1 {
            const EXPECTED_LINE: &str = "Expected to include: <green>{}<r>\n";
            const SIGNATURE: &str = get_signature("toIncludeRepeated", "<green>expected<r>", false);
            this.throw(
                global,
                SIGNATURE,
                format_args!(
                    concat!("\n\n", "Expected to include: <green>{}<r>\n", "Received: <red>{}<r>\n"),
                    substring_fmt,
                    expect_string_fmt
                ),
            )
        } else {
            const EXPECTED_LINE: &str = "Expected to include: <green>{}<r> <green>{}<r> times \n";
            const SIGNATURE: &str = get_signature("toIncludeRepeated", "<green>expected<r>", false);
            this.throw(
                global,
                SIGNATURE,
                format_args!(
                    concat!("\n\n", "Expected to include: <green>{}<r> <green>{}<r> times \n", "Received: <red>{}<r>\n"),
                    substring_fmt,
                    times_fmt,
                    expect_string_fmt
                ),
            )
        }
    }
}

// ──────────────────────────────────────────────────────────────────────────
// PORT STATUS
//   source:     src/test_runner/expect/toIncludeRepeated.zig (110 lines)
//   confidence: medium
//   todos:      1
//   notes:      scopeguard on post_match aliases &mut Self; Formatter borrow shape and Expect::throw fmt-arg signature need Phase B attention
// ──────────────────────────────────────────────────────────────────────────
